### PR TITLE
(RES): introduce mod_decl references

### PR DESCRIPTION
**Pull request by matklad**
Module declaration like `mod foo;` is also a reference. We cannot reuse
existing `RustReferenceImpl` here, because it is tied to paths, and we
don't have a path here. So let's create a separate
`RustModDeclReferenceImpl` for this use case.

As a collateral damage, `RustReferenceImpl` is renamed to
`RustPathReferenceImpl` and is fixed: it erroneously used text range in
the document, instead of a text range relative to a parent.

@atsky, @alexeykudinkin please review

### CODE REVIEW

**Comment by atsky** [comment]
Link on whole mod_decl_item looks strange.
osxdaily

**Comment by matklad** [comment]
@atsky fixed

**Comment by matklad** [comment]
Anything else?

File: src/org/rust/lang/core/resolve/ref/RustModReferenceImpl.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.rust.lang.core.resolve.ref
+
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiReferenceBase
+import org.rust.lang.core.psi.RustModDeclItem
+import org.rust.lang.core.psi.RustModItem
+import org.rust.lang.core.psi.util.moduleFile
+
+class RustModReferenceImpl(modDecl: RustModDeclItem) :
+        PsiReferenceBase<RustModDeclItem>(modDecl, modDecl.identifierRange, /* soft = */ false)
```
Comment (alexeykudinkin) [inline src/org/rust/lang/core/resolve/ref/RustModReferenceImpl.kt]: Style should be consistent.

Comment (matklad) [inline src/org/rust/lang/core/resolve/ref/RustModReferenceImpl.kt]: Is there any way I can convince IDEA to indent `:` if it is on the new line? Currently, it autoformats to 

```Kotlin
class RustModReferenceImpl(modDecl: RustModDeclItem)
: PsiReferenceBase<RustModDeclItem>(modDecl, modDecl.identifierRange, /* soft = */ false)
  , RustReference {
```

**Comment by alexeykudinkin** [comment]
`RustQualifiedReferenceImpl` is more suitable than`RustPathReferenceImpl`: it has nothing to do with paths.

File: src/org/rust/lang/core/resolve/ref/RustModReferenceImpl.kt (hunk 1 above)
Comment (alexeykudinkin) [inline src/org/rust/lang/core/resolve/ref/RustModReferenceImpl.kt]: IDEA's formatter unfortunately isn't that flexible.

**Comment by matklad** [comment]
>RustQualifiedReferenceImpl is more suitable thanRustPathReferenceImpl: it has nothing to do with paths.

fixed

**Comment by matklad** [comment]
Если критических замечаний нет, то я вмерджу, чтобы дальше use resolve пилить

**Comment by alexeykudinkin** [comment]
LGTM

**Pull request merged**